src/test/kotlin/org/jetbrains/kotlinx/jupyter/test/repl/ReplVarsTest.kt

```diff
@@ -318,4 +321,121 @@ class ReplVarsTest : AbstractSingleReplTest() {
     }
+
+    @Test
+    fun testAnonymousObjectRendering() {
+        eval("42")
+        eval("val sim = object : ArrayList<String>() {}")
+        val res = eval("sim").resultValue
+        res.toString() shouldBe "[]"
+    }
+
+    @Test
+    fun testOutVarRendering() {
+        eval("Out").resultValue.shouldNotBeNull()
+    }
+
+    @Test
+    fun testProperBiRecursionHandling() {
+        eval(
+            """
+            val l = mutableListOf<Any>()
+            l.add(listOf(l))
+            
+            val m = mutableMapOf<Int, Any>(1 to l)
+            
+            val z = setOf(1, 2, 4)
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        var state = repl.notebook.variablesState
+        state["l"]!!.stringValue shouldBe "ArrayList: [exception thrown: java.lang.StackOverflowError]"
+        state["m"]!!.stringValue shouldBe "LinkedHashMap: [exception thrown: java.lang.StackOverflowError]"
+        eval(
+            """
+            val m = mutableMapOf<Int, Any>(1 to "abc")
+            """.trimIndent(),
+            jupyterId = 2
+        )
+        state = repl.notebook.variablesState
+        state["l"]!!.stringValue shouldBe "ArrayList: [exception thrown: java.lang.StackOverflowError]"
+        state["m"]!!.stringValue shouldNotBe "LinkedHashMap: [exception thrown: java.lang.StackOverflowError]"
+    }
+
+    @Test
+    fun testUnchangedVars() {
+        eval(
+            """
+            var l = 11111
+            val m = "abc"
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        eval(
+            """
+            l += 11111
+            """.trimIndent(),
+            jupyterId = 2
+        ).metadata.evaluatedVariablesState
+        val state: Set<String> = repl.notebook.unchangedVariables
+        state.size.shouldBe(1)
+        state.contains("m").shouldBe(true)
+    }
+
+    @Test
+    fun testMutableList() {
+        eval(
+            """
+            val l = mutableListOf(1, 2, 3, 4)
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        val serializer = repl.variablesSerializer
+        val res = eval(
+            """
+            l.add(5)
+            """.trimIndent(),
+            jupyterId = 2
+        ).metadata.evaluatedVariablesState
+        val innerList = res["l"]!!.fieldDescriptor["elementData"]!!.fieldDescriptor["data"]
+        val newData = serializer.doIncrementalSerialization(0, "l", "data", innerList!!)
+        newData.isContainer shouldBe true
+        // since there might be null placeholders in array after addition
+        newData.fieldDescriptor.size shouldBeGreaterThanOrEqual 5
+    }
+
+    @Test
+    fun unchangedVariablesGapedRedefinition() {
+        eval(
+            """
+            private val x = "abcd"
+            var f = 47
+            internal val z = 47
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        var state = repl.notebook.unchangedVariables
+        state.size.shouldBe(3)
```

Comment:
Actually, what do you test here? What variables should be contained in this list? What if this test fail, how will you understand what variables have changed without debugging?